File: frontend/src/chapters/ch18/ch18_15_ex.ts
import { Lesson } from '../../courses';

export const ch18_15_ex: Lesson = {
    id: 'ch18-15-ex',
    title: 'Bài tập 18.15: Volatile Memory Access',
    duration: '15 phút',
    type: 'practice',
    isExercise: true,
    content: `<p>Hãy tìm hiểu về volatile memory!</p>`,
    defaultCode: `fn main() {
    let data = vec![1i32, 2, 3, 4, 5];
    let ptr = data.as_ptr();
    unsafe { println!("{}", *ptr); }
}
`,
    expectedOutput: '1',
    testCases: [{ input: '*ptr', expectedOutput: '1', description: 'Volatile works' }]
};

File: backend/src/exercises/ch18/ch18_15_ex.rs
// Exercise - ch18_15_ex: Volatile Memory Access


fn main() {
    let data = vec![1i32, 2, 3, 4, 5];
    let ptr = data.as_ptr();

    unsafe {
        for i in 0..data.len() {
            println!("Value at {}: {}", i, *ptr.offset(i as isize));
        }
    }
}
